### ISSUE TITLE
RNBGL not tracking in the background in a timely manner

### ISSUE THREAD

**Issue by bengtan** [transistorsoft/react-native-background-geolocation#778]
(RNBGL = react-native-background-geolocation)

## Your Environment

I'm using the demo app BGGeolocation at commit point https://github.com/transistorsoft/rn-background-geolocation-demo/commit/f398f9b4 and compiled with XCode 10.1.

* Plugin version: 3.0.9
* Platform: iOS
* OS version: 12.3.1
* Device manufacturer / model: iPhone 6
* React Native version: Whatever the demo uses
* Plugin config: See logs

## Expected Behavior

After I drive for more than 200m, I expect RNBGL to start uploading location data points in the background.

## Actual Behavior

It didn't. Please see below.

## Steps to Reproduce

* Load the BGGeolocation app onto a phone.
* Open the 'Advanced' sub-app and enable it/tracking.
* Press the phone's Home key so as to put BGGeolocation in the background.
* Drive for about 1500 metres.

(I didn't change any default settings that I know of.)

## Context, Description, Discussion

I've been trying to track down a problem in our app, and I decided to try your demo app to see if I could isolate the problem. I observed some behaviour with your demo app which I didn't expect. Not sure if there is some bug somewhere, or maybe I misunderstand something.

(Local time is GMT +8)

### Test set 'bengtan1'

https://tracker.transistorsoft.com/bengtan1

Two locations: A = Trevor Gribble Park, Bull Creek. B = Harry Buckley Park, Bateman.
Date: 20190722

First leg:

* Started at location A. Started driving to location B.
* 14:58pm (06:58 UTC) Arrived at location B. Stopped driving. 
  * Noticed there were no new location data points uploaded during the drive. Intentionally wait a few minutes.
* 15:01 Press home button on the Phone. Wait a few seconds. Still no new data points.
  * Unlock the phone and opened BGGeolocation. I saw the BGGeolocation splash screen (suggesting that it was being restarted). A few data points got uploaded (of location B).

Second leg:

* 15:08 Leave location B and start driving to location A.
* 15:13 Arrived at location A. Stopped driving. 
  * Noticed there were no new location data points uploaded during the drive. 
  * Several seconds after stopping, a few location data points got uploaded (BGGeolocation started making some noises.)

### Test set 'bengtan2'

https://tracker.transistorsoft.com/bengtan2

First leg:

* Started at location A. Started driving to location B.
* 15:21 Arrived at location B. No location updates during the drive.
* 15:24 Press home button. New location data points got uploaded immediately.

Second leg:

* 15:28 Leave location B and drive to location A.
* 15:32 Arrived at location A. No location updates during the drive.
* 15:35 Press home button. New location data points got uploaded immediately.

### Notes

When I say 'press home button', I didn't screen-unlock the phone. I just pressed the Home button so the screen would light up (for a few seconds).

Locations A and B are about 1500 metres apart. That should be far enough to satisfy the requirement: (Exiting the stationary-geofence typically requires ~200 meters of movement)[https://github.com/transistorsoft/react-native-background-geolocation/wiki/Philosophy-of-Operation]

It's important not to press the screen or unlock the phone during the drive.

### Discussion

Test set bengtan1 was just odd. Random crash/failure-to-start and then a random late 'waking up'.

Test set bengtan2 ... it seems like iOS wants to do some battery optimisation and didn't let the app run unless something else (ie. potential screen unlock) was also running?

Most of the time I test, I run into problems. I rarely get it working (in the background) at all.

I dunno what's happening. It's weird.

I tested RNBGL last year and it worked fine. Albeit that was the 2.x branch. Maybe the 3.x branch is very different and has a few bugs? I hope we don't have to downgrade to 2.x.

## Debug logs

See attached

[background-geolocation.log.gz](https://github.com/transistorsoft/react-native-background-geolocation/files/3417173/background-geolocation.log.gz)

**Comment by bengtan** [comment]
Updated my iPhone to iOS 12.4 today and re-tested. Same results.

### Test set 'bengtan3'

I chose a different location for location A, but the results are the same.

* 2:01pm (06:01am UTC) At Location A. Started driving
* 2:05 Arrived at location B. No location data points.
* 2:08 Press home button on iPhone. New location data points got uploaded immediately.

* 2:11 Leave location B
* 2:16 Arrive at location A. No location data points.
* 2:19 Press home button on iPhone. New location data points got uploaded immediately.

### Test set 'bengtan4'

2:22pm At location A. Started driving
2:27 Arrived at location B. No location data points.
2:30 Press home button on iPhone. New location data points got uploaded immediately.

2:34 Leave location B
2:38 Arrive at location A. No location data points.
2:41 Press home button on iPhone. New location data points got uploaded immediately.

----

Also, this problem seems to be related to the stationary->moving transition.

If I'm in the middle of driving and I press Home on the iPhone, BGGeolocation wakes up and starts tracking ... and keeps tracking whilst I am still driving (I was driving for about an additional 5 ish minutes)

See log below

[background-geolocation.log.gz](https://github.com/transistorsoft/react-native-background-geolocation/files/3420650/background-geolocation.log.gz)

**Comment by christocracy** [comment]
>Updated my iPhone to iOS 12.4 today and re-tested. Same results.

After upgrading iOS, always reboot your phone.

**Comment by bengtan** [comment]
> After upgrading iOS, always reboot your phone.

Okay, I rebooted and did some more testing. I'm not sure if rebooting made a difference, but anyway ...

I realised that my tests sets above are flawed because I had a mobile hotspot with me during the tests and (probably) the phone was connecting to the mobile hotspot's wifi, making the tests unrepresentative.

So, please ignore all previous test sets and let's start again.

The following all happens after a reboot.

### Test set 'bengtan7'

Date: 20190724

Leg 1

13:51pm Leave location A
13:55 Arrive at location B. Some location data points got uploaded mid-way

Leg 2

13:59 Leave location B
14:03 Arrive at location A. Some location data points got uploaded mid-way

So this test set actually 'mostly' worked. I say mostly because RNBGL only woke up after travelling about 1000m, instead of 200m as per Apple's docs, but I'll let that slide for now.

### Test set 'bengtan8'

Leg 1

14:07pm Leave location A
14:11 Arrive at location B. No location data points got uploaded.
14:14 Press home key. Phone screen lighted up for a few seconds. No location data points got uploaded.

Odd. Very odd. I waited a few seconds and then stopped waiting and tried to open BGGeolocation.

I saw the BGGeolocation splash screen for a few seconds and then it went back to the background (ie. crashed). Tried it a couple more times with same result. Tried swipe-killing it and restart. Same result.

14:22 Tried to restart BGGeolocation. Still kept crashing.

After this, I did some more driving in order to test our app.

??:?? More driving around.
14:56-ish BGGeolocation started beeping. It had sucessfully restarted itself and woken up.

If I look in the log, I see: 

```
2019-07-24 14:04:09.522 🔵-[TSLocationManager onSuspend:] enabled? 1)

2019-07-24 14:56:54.590 ℹ️-[TSLocationManager init] {
```

During this time, I had tried several times to restart BGGeolocation. If it did get woken up on the OS, looks like it kept crashing.

So I dunno if the crash is due to the RNBGL library or due to BGGeolocation app code.

[background-geolocation.log.gz](https://github.com/transistorsoft/react-native-background-geolocation/files/3425551/background-geolocation.log.gz)

**Comment by christocracy** [comment]
>only woke up after travelling about 1000m, instead of 200m as per Apple's docs

200 meters is not a guarantee.  It's only typical.  It could take 1000 meters or more depending on the nature of the network environment.

Here's a recent road-trip I did last week (July 18).  The green polylines are where a state-change occurred from *stationary->moving*:
![](https://dl.dropbox.com/s/131lpytl29ntosq/Screenshot%202019-07-24%2009.50.18.png?dl=1)

Now look what happens later that day after exiting a relatively remote campground in a rural area with poor cell-service, where it took several km to trigger:

![](https://dl.dropbox.com/s/odjoecp9jg6nwrl/Screenshot%202019-07-24%2009.52.48.png?dl=1)
- Zoomed out:
![](https://dl.dropbox.com/s/lywnnnwgfk4sv1r/Screenshot%202019-07-24%2009.53.30.png?dl=1)

**Comment by christocracy** [comment]
The last location recorded before your app was suspended at `14:04:09` ([map](https://www.google.com/maps/place/32%C2%B003'23.3%22N+115%C2%B052'13.5%22E/@32.0558814,115.8663564,1969m/data=!3m1!1e3!4m5!3m4!1s0x0:0x0!8m2!3d32.0564811!4d115.8704265)):
```
2019-07-24 14:04:04.770 
📍<-32.05648110,+115.87042646> +/- 10.00m (speed 0.00 mps / course 179.65) @ 24/07/2019, 2:04:03 PM Australian Western Standard Time
```
This looks like a relatively rural location.  The device is going to need a cell-tower change before it wakes up decides to switch out of the *stationary* state.

At `14:56:56`, your app re-awakens here ([map](https://www.google.com/maps/place/32%C2%B003'17.4%22N+115%C2%B052'21.3%22E/@32.0548503,115.8703817,987m/data=!3m1!1e3!4m5!3m4!1s0x0:0x0!8m2!3d32.0548458!4d115.8725757))
```
2019-07-24 14:56:55.968 
📍<-32.05484583,+115.87257567> +/- 5.00m (speed 15.65 mps / course 0.16) @ 24/07/2019, 2:56:54 PM Australian Western Standard Time
```

The plugin enters the *moving* state and begins tracking for several minutes before you tell the plugin to `#stop`:
```
2019-07-24 14:57:58.753 
╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager stop] 
╚═══════════════════════════════════════════════════════════
```

**Comment by christocracy** [comment]
I suggest you go for a much longer road-trip of >10km.

**Comment by bengtan** [comment]
> This looks like a relatively rural location. 
> ...
> your app re-awakens here (map)

Uh, no. Your link points to 32.05484583,115.87257567.

The coordinates in the log is -32.05484583,115.87257567: https://www.google.com/maps/search/-32.05484583,115.87257567

You're looking at the wrong place. If you look at the correct place, it's very much urban up with lots of roads and stuff.

> The plugin enters the moving state and begins tracking for several minutes before you tell the plugin to #stop:

Not me. I'm not telling it to stop. I'm using your RNBGL demo app BGGeolocation. It's all your code.

----

Furthermore, it doesn't explain why I wasn't able to restart BGGeolocation during that 50 minute period despite repeated attempts to open it (including a couple of double-tap-home-swipe-kills).

BTW, I'd be happy to run a debug version of your BGGeolocation (with plugin) over here, if you wanna try debug the crash.

----

> I suggest you go for a much longer road-trip of >10km.

As mentioned above, I think you're making this suggestion based on a mistaken assumption arising from you looking at the wrong place.

**Comment by christocracy** [comment]
I missed the prefixed `-` (negative) while copying the lat/lng.  I've updated my previous post.

**Comment by christocracy** [comment]
When you see this, *you* are executing the `#stop` method.
```
2019-07-24 14:57:58.753 
╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager stop] 
╚═══════════════════════════════════════════════════════════
```

**Comment by christocracy** [comment]
In fact, you execute `#stop` *3 times* within a second:
```
2019-07-24 14:57:58.753 
╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager stop] 
╚═══════════════════════════════════════════════════════════
.
.
.
2019-07-24 14:58:03.486 
╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager stop] 
╚═══════════════════════════════════════════════════════════
.
.
.
2019-07-24 14:58:12.309 
╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager stop] 
╚═══════════════════════════════════════════════════════════
```

**Comment by bengtan** [comment]
> When you see this, you are executing the #stop method.
>
> 2019-07-24 14:57:58.753 
> ╔═══════════════════════════════════════════════════════════
> ║ -[TSLocationManager stop] 
> ╚═══════════════════════════════════════════════════════════

I did disable tracking. That's because I was finished testing for the day so I turned it off.

But the (thing that I think is a) 'crash' happened before that

```
2019-07-24 14:04:09.522 🔵-[TSLocationManager onSuspend:] enabled? 1)

2019-07-24 14:56:54.590 ℹ️-[TSLocationManager init] {
```

After the crash, BGGeolocation eventually wakes up and seems to have fixed itself. Then, I turned it off. I'm not interested in what happened after it woke up. That all looks fine to me.

It's what happened beforehand that is ... unexplained.

**Comment by christocracy** [comment]
If you suspect your iOS app is crashing, why don't you plug your device into usb and observe your device's crash logs in XCode?

**Comment by bengtan** [comment]
> If you suspect your iOS app

Please stop saying it's 'my' app. I'm testing using your demo app BGGeolocation. There is none of my code involved.

(Yes, once, above, I did say that I continued testing 'my' app afterwards, but that was just a general comment. All the technical stuff I'm posting ie. logs, timestamps, is from BGGeolocation.)

> If you suspect your iOS app is crashing, why don't you plug your device into usb and observe your device's crash logs in XCode?

I don't know if it's crashing. Maybe 'crash' is not the right word. But it's something unexplained, and I'm asking you for help since you're the expert.

So let me rephrase it:

For some reason, BGGeolocation suddenly stopped working between 14:04:09.522 and 14:56:54.590 and I have no idea why. Are you able to offer any comments or ideas?

**Comment by christocracy** [comment]
- What do you see in the XCode crash logs?
- Go for a long field-test >10km.

**Comment by bengtan** [comment]
> What do you see in the XCode crash logs?

Alright, I'll give it a try tomorrow (It's night time over here.) ... and see if I can get anything interesting from the XCode logs.

(Although ... my phone is now on iOS 12.4 and I'm not sure if I can find XCode device support files for 12.4 yet. Just a FYI.)

**Comment by christocracy** [comment]
XCode 10.3 was just released with support for `12.4`:
![](https://dl.dropbox.com/s/x2ea520fyapoah6/Screenshot%202019-07-24%2011.16.25.png?dl=1)

**Comment by bengtan** [comment]
I haven't reproduced the crash yet today, but I did find something.

Turns out iPhone's keep crash logs on the phone. Goto Settings » Privacy » Analytics » Analytics Data (on iOS 12).

I'm attaching three crash reports I found. They don't correspond to the period between 14:04:09.522 and 14:56:54.590 (Surprisingly, there's no crash report around that time) but you might still find them interesting.

It looks like BGGeolocation sometimes takes too long to initialise when starting in the background. However, it's not clear whether the problem is in RNGBL, BGGeolocation app code, or somewhere else. It may even be some new React Native initialisation thing.

[BGGeolocation-2019-07-24-114508.ips.txt](https://github.com/transistorsoft/react-native-background-geolocation/files/3429938/BGGeolocation-2019-07-24-114508.ips.txt)

[BGGeolocation-2019-07-24-114524.ips.txt](https://github.com/transistorsoft/react-native-background-geolocation/files/3429939/BGGeolocation-2019-07-24-114524.ips.txt)

[BGGeolocation-2019-07-24-114544.ips.txt](https://github.com/transistorsoft/react-native-background-geolocation/files/3429940/BGGeolocation-2019-07-24-114544.ips.txt)

**Comment by bengtan** [comment]
Okay, got some more information for you. There's about 20 BGGeolocation crash logs on my phone. Looking through them all, there are two types.

The majority are of this type (TYPE 1): 

```
Exception Type:  EXC_CRASH (SIGKILL)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Termination Reason: Namespace SPRINGBOARD, Code 0x8badf00d
Termination Description: SPRINGBOARD, scene-create watchdog transgression: com.transistorsoft.backgroundgeolocation.react exhausted real (wall clock) time allowance of 19.54 seconds | ProcessVisibility: Background | ProcessState: Running | WatchdogEvent: scene-create | WatchdogVisibility: Foreground | WatchdogCPUStatistics: ( | "Elapsed total CPU time (seconds): 0.870 (user 0.870, system 0.000), 36% CPU", | "Elapsed application CPU time (seconds): 0.001, 0% CPU" | )
Triggered by Thread:  0

Thread 0 name:  Dispatch queue: com.apple.main-thread
Thread 0 Crashed:
0   libsystem_kernel.dylib        	0x000000019483c130 semaphore_wait_trap + 8
1   libdispatch.dylib             	0x000000019468b7a0 _dispatch_sema4_wait$VARIANT$mp + 24
2   libdispatch.dylib             	0x000000019468c220 _dispatch_semaphore_wait_slow + 140
3   BGGeolocation                 	0x00000001002b0fd4 -[RCTBundleURLProvider isPackagerRunning:] + 888788 (RCTBundleURLProvider.m:94)
4   BGGeolocation                 	0x00000001002b1668 -[RCTBundleURLProvider guessPackagerHost] + 890472 (RCTBundleURLProvider.m:109)
5   BGGeolocation                 	0x00000001002b1878 -[RCTBundleURLProvider packagerServerHost] + 891000 (RCTBundleURLProvider.m:123)
6   BGGeolocation                 	0x00000001002b196c -[RCTBundleURLProvider jsBundleURLForBundleRoot:fallbackResource:fallbackExtension:] + 891244 (RCTBundleURLProvider.m:133)
7   BGGeolocation                 	0x00000001002b1b0c -[RCTBundleURLProvider jsBundleURLForBundleRoot:fallbackResource:] + 891660 (RCTBundleURLProvider.m:146)
8   BGGeolocation                 	0x00000001001dd6a0 -[AppDelegate sourceURLForBridge:] + 22176 (AppDelegate.m:40)
9   BGGeolocation                 	0x00000001002af708 -[RCTBridge setUp] + 882440 (RCTBridge.m:336)
10  BGGeolocation                 	0x00000001002aea9c -[RCTBridge initWithDelegate:bundleURL:moduleProvider:launchOptions:] + 879260 (RCTBridge.m:219)
```

Log: [BGGeolocation  24-7-19, 11-45.crash.txt](https://github.com/transistorsoft/react-native-background-geolocation/files/3430615/BGGeolocation.24-7-19.11-45.crash.txt)

The minority (only just a few) are of this type (TYPE 2):

```
Exception Type:  EXC_CRASH (SIGKILL)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Termination Reason: Namespace ASSERTIOND, Code 0x8badf00d
Triggered by Thread:  0

Thread 0 name:  Dispatch queue: com.apple.main-thread
Thread 0 Crashed:
0   libsystem_kernel.dylib        	0x000000019483c0f4 mach_msg_trap + 8
1   libsystem_kernel.dylib        	0x000000019483b5a0 mach_msg + 72
2   CoreFoundation                	0x0000000194c3c120 __CFRunLoopServiceMachPort + 236
3   CoreFoundation                	0x0000000194c37030 __CFRunLoopRun + 1360
4   CoreFoundation                	0x0000000194c367c0 CFRunLoopRunSpecific + 436
5   GraphicsServices              	0x0000000196e3779c GSEventRunModal + 104
6   UIKitCore                     	0x00000001c15b5c38 UIApplicationMain + 212
7   BGGeolocation                 	0x0000000100d7580c main + 22540 (main.m:14)
8   libdyld.dylib                 	0x00000001946fa8e0 start + 4
```

Log: 

[BGGeolocation  24-7-19, 11-38.crash.txt](https://github.com/transistorsoft/react-native-background-geolocation/files/3430618/BGGeolocation.24-7-19.11-38.crash.txt)

Type 1 looks like a RN issue. It's getting killed in the background because it's waiting for the RN packager? Or something.

Type 2 ... I dunno. I'm not familiar with native iOS code. But it's relatively rare, and I'm not sure I care about it much.

**Comment by bengtan** [comment]
With the crashes accounted for, I'm satisfied with the state of this discussion. Looks like RNBGL is not at fault here.

Closing.
